docs/architecture/psa-storage-resilience.md

```diff
@@ -0,0 +1,501 @@
+# PSA storage resilience design
+
+## Introduction
+
+The PSA crypto subsystem includes a persistent key store. It is possible to create a persistent key and read it back later. This must work even if the underlying storage exhibits non-nominal behavior. In this document, _resilience_ means correct behavior of the key store even under if the underlying platform behaves in a non-nominal, but still partially controlled way.
+
+At this point, we are only concerned about one specific form of resilience: to a system crash or power loss. That is, we assume that the underlying platform behaves nominally, except that occasionally it may restart. In the field, this can happen due to a sudden loss of power.
+
+This document explores the problem space, defines a library design and a test design.
+
+## Resilience goals for API functions
+
+**Goal: PSA Crypto API functions are atomic and committing.**
+
+_Atomic_ means that when an application calls an API function, as far as the application is concerned, at any given point in time, the system is either in a state where the function has not started yet, or in a state where the function has returned. The application never needs to worry about an intermediate state.
+
+_Committing_ means that when a function returns, the data has been written to the persistent storage. As a consequence, if the system restarts during a sequence of storage modifications $M_1, M_2, \ldots, M_n$, we know that when the system restarts, a prefix of the sequence has been performed. For example, there will never be a situation where $M_2$ has been performed but not $M_1$.
```

Comment:
> If we assume that every call is commiting, then there's no sequence possible, since the restart will happen before/after a single call.

I don't understand. If `f` and `g` are both committing, you can sequence them as `f(); g();` and you know that if you've observed the system in the intermediate state then it won't revert to the original state, and if you've observed the system in the post-g state then it won't revert to the intermediate or original state state.

Sequencing is difficult if operations are _not_ committing. For example, on a typical desktop/server OS, file writes are not committing, so `write(file1, data1); write(file2, data2);` followed by a crash might end up having written to `file2` but not to `file1`. You need a `sync` call to add a commitment point.

> We're not talking about multithreading here either?

Atomicity is a relevant concept for multithreading as well, but that's not the concern here. Or you can think of it in the context of one thread being your program, and another thread choosing to reboot the system from time to time. An API function is atomic if it doesn't matter that the system reboots while it's running.

Reply:
I was just wondering why this document talks about sequencing, if it's not possible on this, described, layer. It's up to the application to handle sequencing, but from the storage perspective, each call will be atomic. Unless we're talking about multithreading (and one call waiting for the other one to finish updating a file).